### PR TITLE
Integrate imitation - Unify Codebase

### PR DESCRIPTION
**Pull request by patrickab**

### CODE REVIEW

File: prefq/examples/imitation_preference_comparisons.py [hunk 1]
```diff
@@ -91,6 +91,7 @@ def __call__(self) -> Tuple[Sequence[TrajectoryWithRewPair], np.ndarray]:
 
         queries = list(self.pending_queries.values())
         self.pending_queries.clear()
+
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:94]:
```suggestion
```

File: prefq/examples/imitation_preference_comparisons.py [hunk 2]
```diff
@@ -162,7 +163,7 @@ def _get_feedback_from_server(self):
         def _wait_for_feedback_request(url):
             while True:
                 time.sleep(5)
-                response = requests.get(url)
+                response = requests.get(url, timeout=5)
```
Comment (timokau) [inline prefq/examples/imitation_preference_comparisons.py:166]: Minor nit: This change does not belong into the "Unify Codebase for functionality across both examples".

File: prefq/server.py [hunk 3]
```diff
@@ -114,25 +124,39 @@ def receive_feedback():
...
     data = flask.request.json  # Represents incoming client http request in json format
 
     # Extract received JSON data
     is_left_preferred = data["is_left_preferred"]
-    video_filename_left = data["video_filename_left"]
-    video_index_left = int(video_filename_left.split(".")[0]) - 1
-    batch_index = video_index_left // 2
+    left_filename = data["video_filename_left"]
+
+    query_id = left_filename[: -len("-left.webm")]
```
Comment (timokau) [inline prefq/server.py:139]: This is imitation specific. Instead, the client should explicitly add a query ID to the request. The imitation client can extract the ID from the filename, but the server does not need to be aware of this.

File: scripts/run_query_example.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
Comment (timokau) [inline scripts/run_query_example.sh:1]: I think we should rename this, and also rename the `query_client.py`. The name should indicate that this is an example with a static set of videos, so `static_queries.py` and `run_static_queries_example.sh` would be good.

This can be a separate PR though, no need to change anything here.

File: scripts/run_server.sh [hunk 5]
```diff
@@ -3,5 +3,21 @@
 set -e
 set -o pipefail
 
-poetry install
+# We need an older version of setuptools (and pip) to install gym 0.21. Poetry
```
Comment (timokau) [inline scripts/run_server.sh:6]: This should not be committed. While it speeds up local development if you need both scripts, running the server alone does not require gym and users should not be required to install it.

Long-term we should update to `gymnasium` to avoid this workaround.

**Review by timokau — CHANGES_REQUESTED** [review]
Looks pretty good already. I left some minor feedback.

File: scripts/run_query_example.sh (hunk 4 above)
Comment (timokau) [reply scripts/run_query_example.sh:1]: Tracked in #20, closing this.

File: prefq/examples/imitation_preference_comparisons.py (hunk 2 above)
Comment (patrickab) [reply prefq/examples/imitation_preference_comparisons.py:166]: does this hold true, considering, that in query_client.py, the wait_for_feedback() function shows the same behavior? Unifying the Codebase I wanted to enforce the same behavior in both modules.

The precommit checks required a timeout in this module, but not in imitation_preference_comparisons.py (or I might have overseen this change in the interactive rebase & accidently discarted it)

https://github.com/timokau/prefq/pull/19/files#diff-4597eddb0345e4b99f750c587943b1ed1d4ea47f6dc312c1e439fcd48208eed8

File: scripts/run_server.sh (hunk 5 above)
Comment (patrickab) [reply scripts/run_server.sh:6]: what would be a feasible short term solution? how about removing all changes in this file from this pull request, then adding run_server.sh to the .gitignore?

Comment (timokau) [reply scripts/run_server.sh:6]: You can just keep the local change without committing it. So remove all changes from this PR, but you shouldn't add it to the gitignore since changes to that file should not generally be ignored -- just this particular one. I think we can get to #21 pretty soon, so this should not be an issue much longer.

File: prefq/examples/imitation_preference_comparisons.py (hunk 2 above)
Comment (timokau) [reply prefq/examples/imitation_preference_comparisons.py:166]: I thought the commits were about unifying the server code, since the clients don't necessarily need to have the exact same behavior. But you have a point, could be argued that this change does belong here :) Its not very important either way, I'll just close this.

File: prefq/server.py (hunk 3 above)
Comment (patrickab) [reply prefq/server.py:139]: this function refers to the FeedbackClient, not the ImitationClient. Therefore, changes would need to be adjusted in the .html/.js files, & within the server function "serve_video()", Information about the Query ID and the suffix name convention would have to be passed by modifying the default flask.send_from_directory(folder,filename) function, that the server uses, to send videos to the FeedbackClient.

how about instead adding a flask session variables:

app.config["LEFT_FILENAME_SUFFIX"]="-left.webm"
app.config["RIGHT_FILENAME_SUFFIX"]="-right.webm"

and (dis)assemble filenames accordingly? 

this way we can add flexibility without changing the defaults

Comment (timokau) [reply prefq/server.py:139]: As discussed in the meeting, we can send the query ID in the request (omit the filenames) and then have the server save the files using the "-left.webm", "-right.webm" convention. I think we don't need to make this configurable for now. The imitation client can then extract the ID from the filename, the static client (`query_client.py`) can generate an ID from the compared indices (e.g., `"01-04"`).

Comment (patrickab) [reply prefq/server.py:141]: The Query Client now sends Query IDs instead of filenames, then the Server assambles the filenames locally. However, either within the Feedback Client or within the Server Script, the Query ID needs to be extracted from the filename in order to avoid unnessessary overhead.

Comment (timokau) [reply prefq/server.py:141]: Okay, I think relying on the name convention internally is fine since we are in control of the filenames. Thanks for changing this!

**Review by timokau — APPROVED** [review]
Minor feedback: It would have been better to do the renaming of `query_client` in a separate commit or separate PR instead of merging it into the "Unify" commit. Its not important to change that here, just something to keep in mind in the future :)

Looks good overall, thanks a lot!
